patches/server/0928-More-vanilla-friendly-methods-to-update-trades.patch

```diff
@@ -0,0 +1,124 @@
+ 
+@@ -928,7 +928,7 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
+             if (avillagertrades_imerchantrecipeoption != null) {
+                 MerchantOffers merchantrecipelist = this.getOffers();
+ 
+-                this.addOffersFromItemListings(merchantrecipelist, avillagertrades_imerchantrecipeoption, 2);
++                this.addOffersFromItemListings(merchantrecipelist, avillagertrades_imerchantrecipeoption, amount); // Paper
+             }
+         }
+     }
+diff --git a/src/main/java/net/minecraft/world/entity/npc/WanderingTrader.java b/src/main/java/net/minecraft/world/entity/npc/WanderingTrader.java
+index af092098cdfc528bd9f5d771ead1b685aa051bee..a8c87f0ed7df687c7571154c1cdb84759b6faed8 100644
+--- a/src/main/java/net/minecraft/world/entity/npc/WanderingTrader.java
++++ b/src/main/java/net/minecraft/world/entity/npc/WanderingTrader.java
+@@ -129,13 +129,20 @@ public class WanderingTrader extends net.minecraft.world.entity.npc.AbstractVill
+ 
+     @Override
+     protected void updateTrades() {
++        // Paper start
++        updateTrades(5);
++    }
++
++    @Override
++    public void updateTrades(int amount) {
++        // Paper end
+         VillagerTrades.ItemListing[] avillagertrades_imerchantrecipeoption = (VillagerTrades.ItemListing[]) VillagerTrades.WANDERING_TRADER_TRADES.get(1);
+         VillagerTrades.ItemListing[] avillagertrades_imerchantrecipeoption1 = (VillagerTrades.ItemListing[]) VillagerTrades.WANDERING_TRADER_TRADES.get(2);
+ 
+         if (avillagertrades_imerchantrecipeoption != null && avillagertrades_imerchantrecipeoption1 != null) {
+             MerchantOffers merchantrecipelist = this.getOffers();
+ 
+-            this.addOffersFromItemListings(merchantrecipelist, avillagertrades_imerchantrecipeoption, 5);
++            this.addOffersFromItemListings(merchantrecipelist, avillagertrades_imerchantrecipeoption, amount); // Paper
+             int i = this.random.nextInt(avillagertrades_imerchantrecipeoption1.length);
+             VillagerTrades.ItemListing villagertrades_imerchantrecipeoption = avillagertrades_imerchantrecipeoption1[i];
+             MerchantOffer merchantrecipe = villagertrades_imerchantrecipeoption.getOffer(this, this.random);
+diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
+index f0b910df1ee471b4d72d97c6197ab14f2854976e..2cc03fe21c879d2cca0c2d5ac85094203fb0709e 100644
+--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
++++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftVillager.java
+@@ -99,6 +99,30 @@ public class CraftVillager extends CraftAbstractVillager implements Villager {
+     }
+ 
+     // Paper start
++    @Override
++    public boolean increaseLevel(int amount) {
++        Preconditions.checkArgument(amount > 0, "Level earned must be positive");
```

Comment:
The setVillagerLevel method throws an IAE if the amount is invalid. I think this should just check that the amount is greater than 0 and it won't set the villager level to above 5. throw an exception if it does instead of just ignoring it.